### PR TITLE
Split third party pull into Quartz child jobs

### PR DESCRIPTION
**Pull request by maallen**
Splits the third party Smartling pull into separate child jobs that can be processed in parallel across nodes in a worker cluster.

### CODE REVIEW

File: webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJob.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  @Autowired MeterRegistry meterRegistry;
+
+  @Value("${l10n.smartlingPullLocale.quartz.schedulerName:" + DEFAULT_SCHEDULER_NAME + "}")
+  String schedulerName;
+
+  @Override
+  public Void call(SmartlingPullBatchFileJobInput input) throws Exception {
+
+    for (RepositoryLocale locale : input.getRepository().getRepositoryLocales()) {
+      if (locale.getParentLocale() == null) {
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJob.java:40]: preferably reuse `RepositoryService#getRepositoryLocalesWithoutRootLocale()`

File: webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJobInput.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import com.box.l10n.mojito.service.thirdparty.ThirdPartyTMSSmartling;
+import com.box.l10n.mojito.service.thirdparty.smartling.SmartlingOptions;
+import com.box.l10n.mojito.smartling.SmartlingClient;
+import java.util.Map;
+
+public class SmartlingPullBatchFileJobInput {
+
+  Long batchNumber;
+  String pluralSeparator;
+  Repository repository;
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJobInput.java]: it sounds like we're passing an entity as job input. It may create all sorts of issue with serialization/deserialization. I think it is better to stick to simple types.

File: webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJobInput.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  String projectId;
+  SmartlingOptions options;
+  Map<String, String> localeMapping;
+  ThirdPartyTMSSmartling.Prefix filePrefix;
+
+  SmartlingClient smartlingClient;
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJobInput.java]: Similarly here, passing client instances does not seem well suited for "job input". I'm even surprised this is working.

File: webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJob.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        continue;
+      }
+
+      SmartlingPullLocaleFileJobInput smartlingPullLocaleFileJobInput =
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJob.java:44]: maybe introduce a function here to make the fanout logic more obvious

File: webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullLocaleFileJobInput.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  Locale locale;
+  String fileName;
+  String pluralSeparator;
+  Repository repository;
+
+  String smartlingLocale;
+
+  SmartlingOptions options;
+
+  SmartlingClient smartlingClient;
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullLocaleFileJobInput.java]: Same as for previous input class

File: webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullTranslationsJob.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            });
+
+    LongStream.range(0, batchesFor(input.getPluralCount(), input.getBatchSize()))
+        .forEach(
+            num -> {
+              SmartlingPullBatchFileJobInput smartlingPullBatchFileJobInput =
+                  new SmartlingPullBatchFileJobInput();
+              smartlingPullBatchFileJobInput.setBatchNumber(num);
+              smartlingPullBatchFileJobInput.setFilePrefix("plural");
+              setupInput(input, smartlingPullBatchFileJobInput);
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullTranslationsJob.java:46]: some duplicated, maybe instantiate `SmartlingPullBatchFileJobInput` in the function and pass the file prefix as param?

File: webapp/src/test/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJobTest.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    english.setBcp47Tag("en");
+    enRepoLocale = new RepositoryLocale();
+    enRepoLocale.setLocale(english);
+    enRepoLocale.setId(1L);
+    Locale french = new Locale();
+    french.setBcp47Tag("fr");
+    RepositoryLocale frRepoLocale = new RepositoryLocale();
+    frRepoLocale.setLocale(french);
+    frRepoLocale.setId(2L);
+    Locale gaelic = new Locale();
```
Comment (aurambaj) [inline webapp/src/test/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJobTest.java:61]: nice!! :)

File: webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJobInput.java (hunk 3 above)
Comment (maallen) [reply webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJobInput.java]: This is from an old commit, head commit uses primitive types as this did not work 😀

File: webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/ThirdPartyTMSSmartling.java [hunk 8]
```diff
@@ -527,186 +526,99 @@ private SmartlingFile uploadTextUnitsToSmartling(
...
-              }
-
-              Long singulars =
-                  singularCount(
-                      repository.getId(),
-                      LOCALE_EN,
+    PollableFuture<Void> pollableFuture =
+        meterRegistry
+            .timer(
+                "SmartlingSync.pull",
```
[the remark this answers is not in the document]
Comment (maallen) [reply webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/ThirdPartyTMSSmartling.java:545]: The json and glossary sync still operate sequentially so I think we should leave this metric as is for now. 

Will we be able to use the execution time of the top level `SmartlingPullTranslationsJob` as a similar metric? I'm assuming Quartz includes the execution time of child jobs within the parents execution?

**Review by hylstonnb — APPROVED** [review]
In general looks good to me, but this PR changed more than a thousand rows. I would suggest to split big changes in small PRs next time.

File: docker/docker-compose-api-worker.yml [hunk 9]
```diff
@@ -2,7 +2,7 @@ version: '2.4'
 name: mojito
 services:
   db:
-    image: mysql:5.7
+    image: mysql:8.0.34
```
Comment (hylstonnb) [inline docker/docker-compose-api-worker.yml:5]: I feel this change shouldn't in this same PR, but have its own PR instead.

File: webapp/package-lock.json [hunk 10]
```diff
@@ -6047,9 +6047,9 @@
       },
       "dependencies": {
         "node-fetch": {
-          "version": "2.6.12",
-          "resolved": "https://registry.npmjs.org/node-fetch/-/node-fetch-2.6.12.tgz",
-          "integrity": "sha512-C/fGU2E8ToujUivIO0H+tpQ6HWo4eEmchoPIoXtxCrVghxdKq+QOHqEZW7tuP3KlV3bC8FRMO5nMCC7Zm1VP6g==",
+          "version": "2.7.0",
```
Comment (hylstonnb) [inline webapp/package-lock.json:6050]: This change as well could be in its own PR, i feel.

Comment (maallen) [reply webapp/package-lock.json:6050]: That shouldn't be there, will remove from the commit it got added by mistake

File: docker/docker-compose-api-worker.yml (hunk 9 above)
Comment (maallen) [reply docker/docker-compose-api-worker.yml:5]: Agree, I'll move this to a separate PR.

Comment (maallen) [reply docker/docker-compose-api-worker.yml]: New PR: https://github.com/pinterest/mojito/pull/33

**Comment by maallen** [comment]
> In general looks good to me, but this PR changed more than a thousand rows. I would suggest to split big changes in small PRs next time.

Agree in general, problem with this PR is that it's a large PR with a single new piece of functionality so it's difficult to separate without causing test failures or having to ignore tests until all functional updates are merged.

File: webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJob.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+              .build();
+
+      quartzPollableTaskScheduler.scheduleJob(quartzJobInfo);
+    }
+
+    return null;
+  }
+
+  private Map<String, String> parseLocaleMapping(String input) {
+    return Optional.ofNullable(localeMappingHelper.getLocaleMapping(Strings.emptyToNull(input)))
```
[the remark this answers is not in the document]
Comment (maallen) [reply webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/smartling/quartz/SmartlingPullBatchFileJob.java]: Turns out `Strings.emptyToNull` was handling the edge case where an empty string is passed as a parameter for locale mappings. I've raised https://github.com/pinterest/mojito/pull/34 to fix this and added a unit test to catch it if it gets removed in future.
